tensorzero-core/src/inference/types/file.rs

```diff
@@ -101,14 +101,17 @@ pub fn serialize_with_file_data<T: Serialize>(value: &T) -> Result<Value, Error>
 }
 
 /// A file for an inference or a datapoint.
-#[derive(Clone, Debug, Deserialize, PartialEq, Serialize, ts_rs::TS)]
+#[derive(Clone, Debug, PartialEq, ts_rs::TS, Serialize)]
+#[serde(tag = "file_type", rename_all = "snake_case")]
 #[ts(export)]
-#[serde(untagged, deny_unknown_fields)]
+// NOTE(shuyangli, 2025-10-21): we're manually implementing Serialize and Deserialize for a while until we're confident
+// that clients are sending us the correct tagged versions. Serialization always produces tagged format, but
+// deserialization accepts both tagged and untagged formats for backwards compatibility.
+// TODO(#4107): Remove this once we're confident that clients are sending us the tagged version.
```

Comment:
We can't remove this until (unless) we migrate the legacy data in the DB.

Reply:
I think we discussed this and because DB only stored the StoredInput and this is only used over the wire, it should be safe?

Reply:
(Generally this also underscores that we need to separate storage formats and wire formats so we can evolve them independently. I don't know if we have any other couplings but cc @virajmehta )